### PR TITLE
Time Range: Copy-paste Time Range

### LINKED ISSUE #76122: TimeRange: Copy & paste time range
**Why is this needed**:

Customers often need to reuse the same time range in multiple places in Grafana. For example:

- after finishing troubleshooting in Explore, go to a dashboard and apply the same range as in Explore
- after narrowing down the data in a dashboard, go to other tab (e.g. with Explore or an App Plugin) and reuse the same range
- etc.

**What would you like to be added**:

It'd be great to have a way to "copy current time range to clipboard" and "paste time range from clipboard". For example there could be two small buttons to perform both actions: one to copy the range to clipboard and the other to paste it.

For example:

https://github.com/grafana/grafana/assets/745532/c599bb76-8ef8-4767-bd6d-dd51a839b2cd

**Who is this feature for?**

Any user who needs to look at the data within the same time range but using different features.

cc @gtk-grafana @the-it @grafana/explore-squad

### PR DESCRIPTION
**Pull request by harisrozajac**
**What is this feature?**

Allows copying and pasting time range from Explore to Dashboards and vice versa using buttons in time range picker and the following shortcuts: 

- `t c` for copying
- `t v` for pasting

**Why do we need this feature?**

Customers often need to reuse the same time range in multiple places in Grafana. For example:

- After finishing troubleshooting in Explore, go to a dashboard and apply the same range as in Explore
- After narrowing down the data in a dashboard, go to another tab (e.g. with Explore or an App Plugin) and reuse the same range

**Demo**

https://github.com/grafana/grafana/assets/58232930/1024026e-5e66-49c5-9fb7-f24baeb2adce

**Which issue(s) does this PR fix?**:

Fixes #76122 

**Special notes for your reviewer:**

Please check that:
- [ ] You can copy and paste with buttons and shortcuts - To do this, you can copy a time range in Explore and paste in Dashboards and vice versa
- [ ] Paste in random in strings to make sure that validation works

### CODE REVIEW

File: public/app/features/explore/state/time.ts
Comment (harisrozajac) [inline public/app/features/explore/state/time.ts:182]: the logic for reading the text, checking if it's valid, and returning a valid range or error can be extracted into a separate function that can be shared between explore and dashboards, but wasn't sure if we should do that. if we decide to do it, i'm sure where to put that function.

File: public/app/features/explore/state/time.ts
Comment (harisrozajac) [inline public/app/features/explore/state/time.ts:200]: Currently updating only the left pane in split mode if panes are not synced. Maybe we can add some type of notification that tells user that, or add something to Explore docs.

**Review by mckn — COMMENTED** [review]
LGTM! I'm not 100% sure about the placement of the copy/past icons in the time picker. Have you involved UX in this?

**Comment by harisrozajac** [comment]
> LGTM! I'm not 100% sure about the placement of the copy/past icons in the time picker. Have you involved UX in this?

@mckn thank you! I've shared the PR with the UX team, and I'll wait for their feedback.

File: public/app/core/components/TimePicker/TimePickerWithHistory.tsx
Comment (harisrozajac) [inline public/app/core/components/TimePicker/TimePickerWithHistory.tsx:38]: when it comes to buttons, I'm only showing a toast when there's an error on paste. however, when using keyboard shortcuts, I show a success toast on copy, so I'm wondering if we should add the same logic when copying successfully on the button click.

Comment (ashharrison90) [inline public/app/core/components/TimePicker/TimePickerWithHistory.tsx:38]: this is _hyper_ specific to the error you're trying to catch. what happens if we want to catch other errors? do we add a new independent prop for each one? 😅 

could we instead make it a more generic `onError` callback, then you pass these messages to the callback from `TimeRangeContent`

File: public/app/core/components/TimePicker/TimePickerWithHistory.tsx
Comment (ashharrison90) [inline public/app/core/components/TimePicker/TimePickerWithHistory.tsx:41]: also these messages should be translated

**Review by ashharrison90 — COMMENTED** [review]
will let someone from UX weigh in on the button placement/styles etc.

generally looks good code-wise. few things to consider, mostly translations

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.test.tsx
Comment (harisrozajac) [reply packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.test.tsx:139]: sure 👍

File: public/app/features/explore/state/time.ts
Comment (ifrost) [reply public/app/features/explore/state/time.ts:182]: Maybe here: https://github.com/grafana/grafana/blob/main/public/app/core/utils/timePicker.ts?

**Review by ifrost — APPROVED** [review]
Looks great! I'd prefer buttons to be inline with the apply button but It's up to UX to make the final call.

One more thing. I'm getting doubled errors in Explore when pasting invalid range via keyboard shortcut:

Screenshot 2024-01-12 at 13 53 06

**Comment by harisrozajac** [comment]
> One more thing. I'm getting doubled errors in Explore when pasting invalid range via keyboard shortcut:

@ifrost it turns out that all time range events in Explore will also trigger dashboard events because dashboards are subscribed to them. I've added context prop to `setupTimeRangeBindings` to differentiate between explore and dashboard events. that way PastTime event will have the context payload and will prevent unnecessary triggers. let me know if I should add this to other time range events

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx
Comment (harisrozajac) [reply packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx]: turns out i needed `aria-label`. i've added translations for all suggested areas

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx
Comment (ashharrison90) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx:200]: as mentioned, i don't think you need `aria-label` if you've specified a `tooltip`

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx
Comment (ashharrison90) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx:209]: as mentioned, i don't think you need `aria-label` if you've specified a `tooltip`

File: public/app/core/components/TimePicker/TimePickerWithHistory.tsx
Comment (ashharrison90) [inline public/app/core/components/TimePicker/TimePickerWithHistory.tsx:42]: 2 things here:
- [we use a maximum of 3 segments for the keys](https://github.com/grafana/grafana/blob/main/contribute/internationalization.md#phrase-id-naming-convention)
- you should pass the error to the translation as a variable so the translators have as much context as possible for how to translate the string. look up how e.g. `expand-folder-button` works

**Review by ashharrison90 — COMMENTED** [review]
almost there, couple of minor things to fix otherwise lgtm 🥳

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx
Comment (harisrozajac) [reply packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeContent.tsx]: ah i didn't meet the first condition, i just looked at the button itself without triggering the tooltip. i've removed the `aria-label` and tested with VoiceOver and it works as expected. thanks for double-checking this! :)

**Comment by harisrozajac** [comment]
@amy-super thank you for your feedback on this! 

I've addressed all the items you requested:

- [ ]  Added the shortcuts to the shortcut modal so that they are documented in-app.
- [ ]  Used `copy.svg` for copy and `clipboard-alt.svg` (for some reason `clipboard-notes` were not available as a valid type in the Button component) for paste. 
- [ ] Put the copy/paste buttons in-line with the Apply time range button, here's the screenshot:
Screenshot 2024-01-17 at 1 36 07 PM

**Review by ashharrison90 — APPROVED** [review]
lgtm 👍 nice work 🚀

**Pull request merged**
